src/http/client.cr

```diff
@@ -583,11 +583,19 @@ class HTTP::Client
   end
 
   private def exec_internal(request)
-    response = exec_internal_single(request)
+    begin
+      response = exec_internal_single(request)
+    rescue e : IO::Error
+      raise e unless e.os_error == Errno::EPIPE
+    end
     return handle_response(response) if response
 
     # Server probably closed the connection, so retry one
-    close
+    begin
+      close
+    rescue e : IO::Error
+      raise e unless e.os_error == Errno::EPIPE
+    end
```

Comment:
I think it might make sense to rescue EPIPE in `close` directly. If the client is closed when the connection is already broken, that should not be an error IMO.